url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/SessionBase.java

```diff
@@ -0,0 +1,529 @@
+            send_bind (socket, pipes [1]);
+        }
+
+        //  Plug in the engine.
+        assert (engine == null);
+        engine = engine_;
+        engine.plug (io_thread, this);
+    }
+    
+    public void detach() 
+    {
+        //  Engine is dead. Let's forget about it.
+        engine = null;
+
+        //  Remove any half-done messages from the pipes.
+        clean_pipes ();
+
+        //  Send the event to the derived class.
+        detached ();
+
+        //  Just in case there's only a delimiter in the pipe.
+        if (pipe != null)
+            pipe.check_read ();
+    }
+    
+    protected void process_term (int linger_)
+    {
+        assert (!pending);
+
+        //  If the termination of the pipe happens before the term command is
+        //  delivered there's nothing much to do. We can proceed with the
+        //  stadard termination immediately.
+        if (pipe == null) {
+            proceed_with_term ();
+            return;
+        }
+
+        pending = true;
+
+        //  If there's finite linger value, delay the termination.
+        //  If linger is infinite (negative) we don't even have to set
+        //  the timer.
+        if (linger_ > 0) {
+            assert (!has_linger_timer);
+            io_object.add_timer (linger_, linger_timer_id);
+            has_linger_timer = true;
+        }
+
+        //  Start pipe termination process. Delay the termination till all messages
+        //  are processed in case the linger time is non-zero.
+        pipe.terminate (linger_ != 0);
+
+        //  TODO: Should this go into pipe_t::terminate ?
+        //  In case there's no engine and there's only delimiter in the
+        //  pipe it wouldn't be ever read. Thus we check for it explicitly.
+        pipe.check_read ();
+    }
+    
+
+    //  Call this function to move on with the delayed process_term.
+    private void proceed_with_term() {
+        //  The pending phase have just ended.
+        pending = false;
+
+        //  Continue with standard termination.
+        super.process_term (0);
+    }
+
+
+// JDK6    @Override
+    public void timer_event(int id_) {
+        
+        //  Linger period expired. We can proceed with termination even though
+        //  there are still pending messages to be sent.
+        assert (id_ == linger_timer_id);
+        has_linger_timer = false;
+
+        //  Ask pipe to terminate even though there may be pending messages in it.
+        assert (pipe != null);
+        pipe.terminate (false);
+    }
+
+
+    private void detached() {
+        //  Transient session self-destructs after peer disconnects.
+        if (!connect) {
+            terminate ();
+            return;
+        }
+
+        //  For delayed connect situations, terminate the pipe
+        //  and reestablish later on
+        if (pipe != null && options.delay_attach_on_connect == 1
+            && addr.protocol () != "pgm" && addr.protocol () != "epgm") {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.